### PR TITLE
Docs: added code comments for everything in the @grafana/runtime package.

### PR DESCRIPTION
**Pull request by mckn**
**What this PR does / why we need it**:
This will give documentation in intellisense (for most editors) and also add documentation to the web based packages sdk reference for grafana/runtime.

**Special notes for your reviewer**:
- Please have a look that spelling and also that I have described the usage of the different APIs correctly.
- Will generate a new version of the docs after this one is merged.
- Would be good to go through this package and see if we really need to export everything that we are exporting. I my option it is good to keep the exposed contract as small as possible to make it easier to manage in the future.

### CODE REVIEW

**Review by ryantxu — APPROVED** [review]
LGTM

**Comment by mckn** [comment]
> LGTM

There are a lot more docs to be added :)

File: packages/grafana-runtime/src/services/AngularLoader.ts [hunk 1]
```diff
@@ -1,19 +1,81 @@
...
+ *   render() {
+ *     return (
+ *       <div ref={element => (this.element = element)} />
+ *     );
+ *   }
+ * }
+ * ```
+ *
+ * @public
+ */
```
Comment (hugohaggmark) [inline packages/grafana-runtime/src/services/AngularLoader.ts:33]:
```suggestion
/**
 * Used to enable rendering of Angular components within a
 * React component without loosing proper typings.
 *
 * @example
 * ```typescript
 * class Component extends PureComponent<Props> {
 *   element: HTMLElement;
 *   angularComponent: AngularComponent;
 *
 *   componentDidMount() {
 *     const template = '' // Angular template here;
 *     const scopeProps = {}; // Angular scope properties here
 *     const loader = getAngularLoader();
 *     this.angularComponent = loader.load(this.element, scopeProps, template);
 *   }
 *
 *   componentWillUnmount() {
 *     if (this.angularComponent) {
 *       this.angularComponent.destroy();
 *     }
 *   }
 *
 *   render() {
 *     return (
 *       <div ref={element => (this.element = element)} />
 *     );
 *   }
 * }
 * ```
 *
 * @public
 */
```

File: packages/grafana-runtime/src/services/LocationSrv.ts [hunk 2]
```diff
@@ -1,36 +1,84 @@
...
-   * Do not change this unless you are the angular router
+   * Used internally to sync the redux state from Angular to make sure that the redux location
+   * state is in sync when navigating using the Angular router.
+   *
+   * @remarks
+   * Do not change this unless you are the Angular router.
+   *
+   * @internal
    */
   routeParams?: UrlQueryMap;
```
Comment (mckn) [inline packages/grafana-runtime/src/services/LocationSrv.ts:32]: note: should probably move this from the public interface and create a separate internal interface to handle the angular sync.

File: packages/grafana-runtime/src/services/templateSrv.ts [hunk 3]
```diff
@@ -1,13 +1,31 @@
 import { VariableModel } from '@grafana/data';
 
+/**
+ * Via the TemplateSrv consumers get access to all the available template variables
+ * that can be used within the current active dashboard.
+ *
+ * For a mor in-depth description visit: https://grafana.com/docs/grafana/latest/reference/templating
+ */
 export interface TemplateSrv {
   getVariables(): VariableModel[];
```
Comment (mckn) [inline packages/grafana-runtime/src/services/templateSrv.ts:10]: note: aren't we missing the important `replace` function here?

File: packages/grafana-runtime/src/services/AngularLoader.ts [hunk 4]
```diff
@@ -1,19 +1,87 @@
...
 export interface AngularComponent {
+  /**
+   * Should be called when the react component will unmount.
```
Comment (dprokop) [inline packages/grafana-runtime/src/services/AngularLoader.ts:36]:
```suggestion
   * Should be called when the React component will unmount.
```

File: packages/grafana-runtime/src/services/AngularLoader.ts [hunk 5]
```diff
@@ -1,19 +1,87 @@
...
+ *
+ * @example
+ * ```typescript
+ * class Component extends PureComponent<Props> {
+ *   element: HTMLElement;
+ *   angularComponent: AngularComponent;
+ *
+ *   componentDidMount() {
+ *     const template = '<angular-component />' // angular template here;
+ *     const scopeProps = { ctrl: angularController }; // angular scope properties here
```
Comment (dprokop) [inline packages/grafana-runtime/src/services/AngularLoader.ts:13]:
```suggestion
 *     const template = '<angular-component />' // Angular template here;
 *     const scopeProps = { ctrl: angularController }; // Angular scope properties here
```

File: packages/grafana-runtime/src/services/AngularLoader.ts [hunk 6]
```diff
@@ -1,19 +1,87 @@
+/**
+ * Used to enable rendering of angular components within a
```
Comment (dprokop) [inline packages/grafana-runtime/src/services/AngularLoader.ts:2]:
```suggestion
 * Used to enable rendering of Angular components within a
```

File: packages/grafana-runtime/src/services/AngularLoader.ts [hunk 7]
```diff
@@ -1,19 +1,87 @@
...
+ * react component without loosing proper typings.
```
Comment (dprokop) [inline packages/grafana-runtime/src/services/AngularLoader.ts:3]:
```suggestion
 * React component without loosing proper typings.
```

File: packages/grafana-runtime/src/services/AngularLoader.ts [hunk 8]
```diff
@@ -1,19 +1,87 @@
...
+ *
+ * @public
+ */
 export interface AngularComponent {
+  /**
+   * Should be called when the react component will unmount.
+   */
   destroy(): void;
+  /**
+   * Can be used to trigger a re-render of the angular component.
```
Comment (dprokop) [inline packages/grafana-runtime/src/services/AngularLoader.ts:40]:
```suggestion
   * Can be used to trigger a re-render of the Angular component.
```

File: packages/grafana-runtime/src/services/AngularLoader.ts [hunk 9]
```diff
@@ -1,19 +1,87 @@
...
+   */
   digest(): void;
+  /**
+   * Used to access the angular scope from the react component.
```
Comment (dprokop) [inline packages/grafana-runtime/src/services/AngularLoader.ts:44]:
```suggestion
   * Used to access the angular scope from the React component.
```

File: packages/grafana-runtime/src/services/AngularLoader.ts [hunk 10]
```diff
@@ -1,19 +1,87 @@
...
+   */
   getScope(): any;
 }
 
+/**
+ * Used to load an angular component from the context of a react component.
```
Comment (dprokop) [inline packages/grafana-runtime/src/services/AngularLoader.ts:50]:
```suggestion
 * Used to load an Angular component from the context of a React component.
```

File: packages/grafana-runtime/src/services/AngularLoader.ts [hunk 11]
```diff
@@ -1,19 +1,87 @@
...
+ * via the the {@link getAngularLoader()} to the rest of the application.
+ *
+ * @internal
+ */
 export function setAngularLoader(v: AngularLoader) {
   instance = v;
 }
 
+/**
+ * Used to retrieve the {@link AngularLoader} that enables the use of angular
```
Comment (dprokop) [inline packages/grafana-runtime/src/services/AngularLoader.ts:78]:
```suggestion
 * Used to retrieve the {@link AngularLoader} that enables the use of Angular
```

File: packages/grafana-runtime/src/services/AngularLoader.ts [hunk 12]
```diff
@@ -1,19 +1,87 @@
...
+ * components within a react component.
```
Comment (dprokop) [inline packages/grafana-runtime/src/services/AngularLoader.ts:79]:
```suggestion
 * components within a React component.
```

File: packages/grafana-runtime/src/services/LocationSrv.ts [hunk 13]
```diff
@@ -1,36 +1,84 @@
...
 
   /**
-   * Add the query argument to the existing URL
+   * If set to true, the query argument will be added to the existing URL.
    */
   partial?: boolean;
 
   /**
-   * Do not change this unless you are the angular router
+   * Used internally to sync the redux state from Angular to make sure that the redux location
```
Comment (dprokop) [inline packages/grafana-runtime/src/services/LocationSrv.ts:24]:
```suggestion
   * Used internally to sync the Redux state from Angular to make sure that the Redux location
```

File: packages/grafana-runtime/src/services/backendSrv.ts [hunk 14]
```diff
@@ -1,50 +1,92 @@
...
+   * new one.
+   */
   requestId?: string;
-
-  // Allow any other parameters
   [key: string]: any;
 };
 
+/**
+ * Used to communicate via http(s) to a remote backend such as the grafana backend,
```
Comment (dprokop) [inline packages/grafana-runtime/src/services/backendSrv.ts:42]:
```suggestion
 * Used to communicate via http(s) to a remote backend such as the Grafana backend,
```

File: packages/grafana-runtime/src/services/backendSrv.ts [hunk 15]
```diff
@@ -1,50 +1,92 @@
...
+ *
+ * @internal
+ */
 export const setBackendSrv = (instance: BackendSrv) => {
   singletonInstance = instance;
 };
 
+/**
+ * Used to retrieve the {@link BackendSrv} that can be used to communicate
+ * via http(s) to a remote backend such as the grafana backend, a datasource etc.
```
Comment (dprokop) [inline packages/grafana-runtime/src/services/backendSrv.ts:88]:
```suggestion
 * via http(s) to a remote backend such as the Grafana backend, a datasource etc.
```

File: packages/grafana-runtime/src/services/dataSourceSrv.ts [hunk 16]
```diff
@@ -1,15 +1,39 @@
 import { ScopedVars, DataSourceApi } from '@grafana/data';
 
+/**
+ * This is the entry point for communicating with a datasource that is added as
+ * a plugin (both external and internal). Via this service you will get access
+ * to the {@link @grafana/data#DataSourceApi | DataSourceApi} that have a rich api for
```
Comment (dprokop) [inline packages/grafana-runtime/src/services/dataSourceSrv.ts:6]:
```suggestion
 * to the {@link @grafana/data#DataSourceApi | DataSourceApi} that have a rich API for
```

**Review by dprokop — APPROVED** [review]
LGTM. Left some suggestions regarding lib names capitalisation

File: packages/grafana-runtime/src/services/EchoSrv.ts [hunk 17]
```diff
@@ -25,40 +40,96 @@ export interface EchoMeta {
...
   return singletonInstance;
 }
 
+/**
+ * Used to register echo backends that will receive Grafana echo events during application
+ * runtime.
+ *
+ * @public
+ */
 export const registerEchoBackend = (backend: EchoBackend) => {
```
Comment (mckn) [inline packages/grafana-runtime/src/services/EchoSrv.ts:133]: note: Feels a bit dangerous to give plugins etc. the possibility to listen to all meta events within Grafana. Do we send any sensitive information?

File: packages/grafana-runtime/src/services/EchoSrv.ts [hunk 18]
```diff
@@ -25,40 +40,96 @@ export interface EchoMeta {
...
   addEvent: (event: T) => void;
 }
 
+/**
+ * Type interface describing an echo event.
+ *
+ * @public
+ */
 export interface EchoEvent<T extends EchoEventType = any, P = any> {
   type: EchoEventType;
```
Comment (mckn) [inline packages/grafana-runtime/src/services/EchoSrv.ts:61]: note: this one should probably not be of type `EchoEventType` but rather `T`.

File: packages/grafana-runtime/src/types/analytics.ts [hunk 19]
```diff
@@ -1,12 +1,19 @@
 import { EchoEvent, EchoEventType } from '../services/EchoSrv';
 
+/**
+ * Describes the basic dashboard information that can be passed as the meta
+ * analytics payload.
+ */
 export interface DashboardInfo {
```
Comment (mckn) [inline packages/grafana-runtime/src/types/analytics.ts:7]: note: I don't really see the reason why we want to export the "base" interfaces outside the runtime package.

File: packages/grafana-runtime/src/utils/plugin.ts [hunk 20]
```diff
@@ -3,13 +3,29 @@ import { config } from '../config';
...
+ */
 export interface PluginCssOptions {
   light: string;
   dark: string;
 }
 
+/**
+ * @internal
+ */
 export const SystemJS = System;
```
Comment (mckn) [inline packages/grafana-runtime/src/utils/plugin.ts:20]: note: feels a bit dangerous to expose this one via the grafana/runtime

Comment (dprokop) [reply packages/grafana-runtime/src/utils/plugin.ts:20]: yeah, we probably could have some higher level API for plugin loader to be able to interact with SystemJS

File: packages/grafana-runtime/src/services/LocationSrv.ts [hunk 21]
```diff
@@ -1,36 +1,84 @@
...
 
   /*
-   * If true this will replace url state (ie cause no new browser history)
+   * If set to true, this will replace URL state (ie. cause no new browser history).
    */
   replace?: boolean;
 }
 
+/**
+ * Type to represent the value of a singel query variable.
```
Comment (tskarhed) [inline packages/grafana-runtime/src/services/LocationSrv.ts:41]:
```suggestion
 * Type to represent the value of a single query variable.
```

File: packages/grafana-runtime/src/services/EchoSrv.ts (hunk 18 above)
Comment (mckn) [reply packages/grafana-runtime/src/services/EchoSrv.ts:65]: Should be fixed in another PR

File: packages/grafana-runtime/src/services/LocationSrv.ts (hunk 2 above)
Comment (dprokop) [reply packages/grafana-runtime/src/services/LocationSrv.ts:32]: Think we can leave this for now. Gonna be gone with React Router migration.

File: packages/grafana-runtime/src/services/EchoSrv.ts (hunk 17 above)
Comment (dprokop) [reply packages/grafana-runtime/src/services/EchoSrv.ts:137]: Good question, worth aligning with Enterprise team.

File: packages/grafana-runtime/src/utils/plugin.ts (hunk 20 above)
Comment (dprokop) [reply packages/grafana-runtime/src/utils/plugin.ts:20]: Can you create issue around this/

File: packages/grafana-runtime/src/services/EchoSrv.ts (hunk 17 above)
Comment (sakjur) [reply packages/grafana-runtime/src/services/EchoSrv.ts:137]: 🤔

1. We're listening to the meta events from inside of Grafana, so we don't need it to be accessible for plugins.
2. The information is personal but not sensitive, it responds to the question "who loaded what panel when and how long time did it take". The possibly sensitive thing is that the `error` field is a free-text with whatever error was returned.

Comment (mckn) [reply packages/grafana-runtime/src/services/EchoSrv.ts:137]: 1. OK ok, that is good! I don't think, with the current design of the packages, we can limit this. Might be something we revisit after 7.0.

2. Super!

@dprokop what do you think?

Comment (dprokop) [reply packages/grafana-runtime/src/services/EchoSrv.ts:137]: With the above information I think this should be OK. Worth revisiting as you mentioned @mckn

**Pull request merged**